pyproject.toml

```diff
@@ -1,3 +1,54 @@
+[project]
+name="spack"
+version="0"
+description="The spack package manager"
+dependencies=[
+  "six",
+  "types-six",
+  "setuptools",
+  "clingo",
+]
+packages = []
+
+[project.optional-dependencies]
+dev = [
+  "pip>=21.3",
+  "pytest",
+  "pytest-xdist",
+  "setuptools",
+  "click==8.0.2",
+  'black==21.12b0',
+  "mypy",
+  "isort",
+  "flake8",
+  "vermin",
+]
+ci = [
+  "pytest-cov",
+  "codecov[toml]",
+]
+
+[build-system]
+requires = ["hatchling"]
+build-backend = "hatchling.build"
```

Comment:
Why hatchling?

Reply:
Honestly, because I couldn't figure out how to get setuptools to just let me do an editable install without erroring out or declaring explicit package parameters.  Poetry, which is my usual weapon of choice, seems to require a good bit of extra boilerplate.   This version made it "just work" so the dependencies will install with pip or hatch, or any pep-517/660 compliant tool presumably, out of the box.  The hatch env stuff I just find really handy, honestly, but that isn't tied to using hatchling like this, if you know a better way to get `pip install` with and without `-e` working to set this up I'm happy to explore alternatives.

Reply:
What version of setuptools were you using? I know there was a recent refactor in v64 that added support for editable installs using `pyproject.toml`. I have no personal preference on build backend, just curious about the decision since I'm never sure which backend I should be using in my own projects.

Reply:
Maintainer here! Hatchling is also now used as the default in the official [Python packaging tutorial](https://packaging.python.org/en/latest/tutorials/packaging-projects/).

Reply:
@ofek do you know how the default was chosen? I haven't yet seen an official comparison between backends and PyPA refuses to admit that "default" == "recommended".

Reply:
I think it was briefly mentioned in https://github.com/spack/spack/pull/31512, and discussed in https://github.com/pypa/packaging.python.org/pull/1031#issuecomment-1124977628

I'll quote Henry who expressed the common sentiment:

> Which backend should be the "only" backend? `tools.setuptools` is still experimental and there are a lot of older pre-PEP 621 copies of it around, and it's large and full of legacy cruft still, causing slower builds and harder error messages - it can be powerful, but it seems to be an overkill for newcomers and simple Python only projects. Flit is lite and fast, with simple error messages, but a crippled file include/exclude system that won't include `LICENSE.md` by default but will include `__pycache__` folders. Hatchling seems to be a nice balance, with reasonable defaults & nice error messages. For a nice newcomer experience, I think Hatchling is likely best.

@trws's experience is common and is by design. I've learned the best and worst parts of other backends in order to implement Hatching, which is why it Just Works ™️ while at the same time being simpler to configure and easier to extend.

Reply:
I don't think there's a way to specify that you want to install to `<prefix>` instead of `<prefix>/lib/pythonX.Y/site-packages`. Some dependencies may also have things in `<prefix>/bin`, so you wouldn't want to map everything to a single directory anyway. @pradyunsg would be the one to ask though.

Reply:
Right, but there are some user-facing scripts and things that really ought to be there.  The other directories like `var` make for more of an interesting question.  If we want users to be able to install multiple spacks in different pythons, then they need to be in `site-packages` but if we want it to work like a normal FHS prefix then there are several things that are more appropriately installed at the root of the prefix.

Reply:
Would `shared-data` work? https://hatch.pypa.io/latest/plugins/builder/wheel/#options

That's what Jupyter uses now https://blog.jupyter.org/packaging-for-jupyter-in-2022-c7be64c38926

Reply:
You can still get the best of both worlds. If things are set up correctly, `<root>/bin/spack*` will be installed to `<prefix>/bin` and `<root>/lib/spack/*` will be installed to `<prefix>/lib/pythonX.Y/site-packages`. Not sure exactly where to put other files, but I believe setuptools/etc. have options for that. This matches the FHS prefix.

Even with normal Python libs, you can't install multiple versions at the same time without having a separate environment for each one. So there's no reason Spack needs to do anything special to support that.

I think what you have right now is the bare minimum that will allow us to `pip install .` in CI. If we want to release on PyPI, we'll need to do quite a bit more. `etc`, `share`, and `var` will need to be installed somewhere, and `opt` may need more thought as well. And since these locations are different in our source code vs at run-time after an install, we'll have to think about how to support both, or think about changing the layout of our source code to match what would be installed.

Reply:
I agree this is nowhere near ready for PyPi, but you might be surprised by how well it actually works as an installed package.  Even with everything just installing to site-packages with a shim for the spack command, everything pretty much just works, because all the paths were relative to the repo before and the relative paths have stayed the same.

Reply:
In setuptools lingo, I think the Spack command is an entry point. Does hatchling also support entry points?

Reply:
It does, and that one works just fine with the new main shim I put in so we could move the path logic out of `bin/spack`.  It's the shell scripts etc. that don't work as entry points.  Seems like shared-data does work though, thanks @ofek!  With that we can actually install to the prefix for some of this stuff, but we might run into path problems, running a test or two. If the path resolution doesn't work this way I'm going to leave everything (but the scripts) in site-packages so things keep working and call it good enough for this PR.

Reply:
In the long run we may actually end up splitting core and packages into two separate libraries so don't worry too much about where var ends up.

Reply:
> Does hatchling also support entry points?

https://hatch.pypa.io/latest/config/metadata/#cli

Reply:
shared-data seems fine, and is the right thing assuming that this isn't code and indeed is random data files. :)

Reply:
@pradyunsg it's a bit of a grey area IMO, it's things like package files, which are technically code but are most definitely not libraries or things meant to be run by users, configuration files, etc.

Reply:
I'm pretty sure that you can remove the bin/spack file, since it's being replaced by the script that will be generated for `project.scripts` declarations. FWIW, those are what the console_scripts entry points from setuptools map to -- I'm quite pleased that we have a friendlier user-facing name for the concept.

Reply:
I think it's a bit of a gray area. We would like to be able to run Spack just by cloning and running `./spack/bin/spack`, but we would also like to be able to install with pip. I think the former use case will actually be more common that the latter, so we can't just get rid of the script.

Reply:
@pradyunsg Yeah, it wouldn't be installed by pip, but we need it there for the "clone and go" use-case.

On installing, the shared-data option definitely works, but splitting our directories up breaks all the relative pathing stuff so things break, really badly.  I did manage to get install to produce a console_script that works, and install all the directories under `site-packages/spack_root` though rather than splatting them all over the place, and the resulting install can install packages and generally kinda just works.

Reply:
To be clear, I wasn't suggesting removing it from the repository, but from the list of files being included in the wheel (`python -m build && unzip -l dist/*.whl` shouldn't list any sort of `.scripts/spack` since that would be overwritten by a launcher script that pip generates instead.